### PR TITLE
Support COMMAND [...] commands

### LINKED ISSUE #2889: Add support for redis command: COMMAND COUNT

### LINKED ISSUE #2890: Add support for redis command: COMMAND DOCS (R7)

### LINKED ISSUE #2891: Add support for redis command: COMMAND INFO

### LINKED ISSUE #2892: Add support for redis command: COMMAND LIST (R7)

### PR DESCRIPTION
**Pull request by Avital-Fine**
Closes #2889, Closes #2890, Closes #2891, Closes #2892

### CODE REVIEW

File: src/main/java/redis/clients/jedis/Protocol.java [hunk 1]
```diff
@@ -236,8 +236,8 @@ public static Object read(final RedisInputStream is) {
     GEORADIUSBYMEMBER, GEORADIUSBYMEMBER_RO, BITFIELD, TOUCH, SWAPDB, MEMORY, BZPOPMIN, BZPOPMAX,
     XADD, XLEN, XDEL, XTRIM, XRANGE, XREVRANGE, XREAD, XACK, XGROUP, XREADGROUP, XPENDING, XCLAIM,
     XAUTOCLAIM, XINFO, BITFIELD_RO, ROLE, FAILOVER, GEOSEARCH, GEOSEARCHSTORE, EVAL_RO, EVALSHA_RO,
-    LOLWUT, EXPIRETIME, PEXPIRETIME, FUNCTION, FCALL, FCALL_RO, LMPOP, BLMPOP, ZMPOP, BZMPOP,
-    @Deprecated STRALGO;
+    LOLWUT, EXPIRETIME, PEXPIRETIME, FUNCTION, FCALL, FCALL_RO, LMPOP, BLMPOP, ZMPOP, BZMPOP,COMMAND,
+    DOCS, GETKEYS, GETKEYSANDFLAGS, @Deprecated STRALGO;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Protocol.java:240]:
```suggestion
    LOLWUT, EXPIRETIME, PEXPIRETIME, FUNCTION, FCALL, FCALL_RO, LMPOP, BLMPOP, ZMPOP, BZMPOP,
    COMMAND, DOCS, GETKEYS, GETKEYSANDFLAGS, @Deprecated STRALGO;
```

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 2]
```diff
@@ -8042,6 +8042,61 @@ public long bitop(final BitOP op, final String destKey, final String... srcKeys)
     return connection.executeCommand(commandObjects.bitop(op, destKey, srcKeys));
   }
 
+  @Override
+  public long commandCount() {
+    checkIsInMultiOrPipeline();
+    connection.sendCommand(COMMAND, COUNT);
+    return connection.getIntegerReply();
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:8049]:
```suggestion
    return connection.executeCommand(commandObjects.commandCount());
```

File: src/main/java/redis/clients/jedis/resps/KeyedFlags.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package redis.clients.jedis.resps;
+
+import java.util.List;
+
+public class KeyedFlags {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/KeyedFlags.java:5]: Remove this. Use `KeyValue<String, List<String>>`.

File: src/main/java/redis/clients/jedis/UnifiedJedis.java [hunk 4]
```diff
@@ -505,6 +505,51 @@ public long dbSize() {
     return executeCommand(commandObjects.keys(pattern));
   }
 
+  @Override
+  public long commandCount() {
+    return executeCommand(commandObjects.commandCount());
+  }
+
+  @Override
+  public List<CommandDocs> commandDocs(String... commands) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:514]:
```suggestion
  public Map<String, CommandDocs> commandDocs(String... commands) {
```

File: src/main/java/redis/clients/jedis/commands/CommandCommands.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   * @return commands list
+   */
+  List<String> commandListFilterByAclcat(String category);
+
+  /**
+   * Return a list of the server's command names filter by glob-like pattern
+   * @param pattern
+   * @return commands list
+   */
+  List<String> commandListFilterByPattern(String pattern);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/CommandCommands.java:72]: We can use a new params class `CommandListFilterByParams`.

File: src/main/java/redis/clients/jedis/resps/CommandDocs.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package redis.clients.jedis.resps;
+
+import java.util.List;
+
+public class CommandDocs {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/CommandDocs.java:5]: `CommandDocument`?

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/Protocol.java [hunk 7]
```diff
@@ -237,7 +237,7 @@ public static Object read(final RedisInputStream is) {
     XADD, XLEN, XDEL, XTRIM, XRANGE, XREVRANGE, XREAD, XACK, XGROUP, XREADGROUP, XPENDING, XCLAIM,
     XAUTOCLAIM, XINFO, BITFIELD_RO, ROLE, FAILOVER, GEOSEARCH, GEOSEARCHSTORE, EVAL_RO, EVALSHA_RO,
     LOLWUT, EXPIRETIME, PEXPIRETIME, FUNCTION, FCALL, FCALL_RO, LMPOP, BLMPOP, ZMPOP, BZMPOP,
-    @Deprecated STRALGO;
+    COMMAND, DOCS, GETKEYS, GETKEYSANDFLAGS, @Deprecated STRALGO;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Protocol.java:240]: Only COMMAND is the command here, others are keywords.

File: src/main/java/redis/clients/jedis/params/CommandListFilterByParams.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  public static CommandListFilterByParams commandListFilterByParams() {
+    return new CommandListFilterByParams();
+  }
+
+  public CommandListFilterByParams filterByModule(String moduleName) {
+    this.moduleName = moduleName;
+    return this;
+  }
+
+  public CommandListFilterByParams filterByAclcat(String category) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/CommandListFilterByParams.java:24]: filterByAclCat

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 9]
```diff
@@ -772,6 +781,72 @@ public String toString() {
...
+        String group = STRING.build(commandData.get(5));
+        String complexity = STRING.build(commandData.get(7));
+        List<String> history = null;
+        if (STRING.build(commandData.get(8)).equals("history")) {
+          List<List<Object>> rawHistory = (List<List<Object>>) commandData.get(9);
+          history = new ArrayList<>(rawHistory.size());
+          for (List<Object> timePoint : rawHistory) {
+            history.add(STRING.build(timePoint.get(0)) + ": " + STRING.build(timePoint.get(1)));
+          }
+        }
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BuilderFactory.java:808]: Create a "single object" builder in CommandDocument class.

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 10]
```diff
@@ -772,6 +781,72 @@ public String toString() {
...
+        List<Object> commandInfo = (List<Object>) rawCommandInfo;
+        String name = STRING.build(commandInfo.get(0));
+        long arity = LONG.build(commandInfo.get(1));
+        List<String> flags = STRING_LIST.build(commandInfo.get(2));
+        long firstKey = LONG.build(commandInfo.get(3));
+        long lastKey = LONG.build(commandInfo.get(4));
+        long step = LONG.build(commandInfo.get(5));
+        List<String> aclCategories = STRING_LIST.build(commandInfo.get(6));
+        List<String> tips = STRING_LIST.build(commandInfo.get(7));
+        List<String> subcommands = STRING_LIST.build(commandInfo.get(9));
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BuilderFactory.java:841]: Create a "single object" builder in CommandInfo class.

File: src/test/java/redis/clients/jedis/commands/jedis/AllKindOfValuesCommandsTest.java [hunk 11]
```diff
@@ -22,25 +22,24 @@
...
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
+
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/commands/jedis/AllKindOfValuesCommandsTest.java:26]:
```suggestion
import java.util.*;
```

File: src/main/java/redis/clients/jedis/commands/JedisCommands.java [hunk 12]
```diff
@@ -2,5 +2,5 @@
 
 public interface JedisCommands extends KeyCommands, StringCommands, ListCommands, HashCommands,
     SetCommands, SortedSetCommands, GeoCommands, HyperLogLogCommands, StreamCommands,
-    ScriptingKeyCommands, FunctionCommands {
+    ScriptingKeyCommands, FunctionCommands, CommandCommands {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/JedisCommands.java:5]:
```suggestion
    ScriptingKeyCommands, FunctionCommands {
```

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 13]
```diff
@@ -3032,6 +3032,35 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
...
+
+  public final CommandObject<List<String>> commandList() {
+    return new CommandObject<>(commandArguments(COMMAND).add(LIST), BuilderFactory.STRING_LIST);
+  }
+
+  public final CommandObject<List<String>> commandListFilterBy(CommandListFilterByParams filterByParams) {
+    return new CommandObject<>(commandArguments(COMMAND).add(LIST).addParams(filterByParams),
+        BuilderFactory.STRING_LIST);
+  }
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:3063]:
```suggestion
```

File: src/main/java/redis/clients/jedis/UnifiedJedis.java [hunk 14]
```diff
@@ -505,6 +505,41 @@ public long dbSize() {
...
+  @Override
+  public List<String> commandList() {
+    return executeCommand(commandObjects.commandList());
+  }
+
+  @Override
+  public List<String> commandListFilterBy(CommandListFilterByParams filterByParams) {
+    return executeCommand(commandObjects.commandListFilterBy(filterByParams));
+  }
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:542]:
```suggestion
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/test/java/redis/clients/jedis/commands/jedis/AllKindOfValuesCommandsTest.java (hunk 11 above)
Comment (Avital-Fine) [reply src/test/java/redis/clients/jedis/commands/jedis/AllKindOfValuesCommandsTest.java:26]: ?

File: src/main/java/redis/clients/jedis/commands/JedisCommands.java (hunk 12 above)
Comment (Avital-Fine) [reply src/main/java/redis/clients/jedis/commands/JedisCommands.java:5]: why?

File: src/main/java/redis/clients/jedis/params/CommandListFilterByParams.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    if (moduleName != null && category == null && pattern == null) {
+      args.add(MODULE);
+      args.add(moduleName);
+    } else if (moduleName == null && category != null && pattern == null) {
+      args.add(ACLCAT);
+      args.add(category);
+    } else if (moduleName == null && category == null && pattern != null) {
+      args.add(PATTERN);
+      args.add(pattern);
+    }
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/CommandListFilterByParams.java:47]: Add and `else` block and throw JedisException.

File: src/main/java/redis/clients/jedis/CommandObjects.java [hunk 16]
```diff
@@ -632,6 +632,29 @@ protected CommandArguments commandArguments(ProtocolCommand command) {
...
+    return new CommandObject<>(commandArguments(COMMAND).add(Keyword.INFO).addObjects((Object[]) commands), BuilderFactory.COMMAND_INFO_RESPONSE);
+  }
+  public final CommandObject<List<String>> commandList() {
+    return new CommandObject<>(commandArguments(COMMAND).add(LIST), BuilderFactory.STRING_LIST);
+  }
+  public final CommandObject<List<String>> commandListFilterBy(CommandListFilterByParams filterByParams) {
+    return new CommandObject<>(commandArguments(COMMAND).add(LIST).addParams(filterByParams),
+            BuilderFactory.STRING_LIST);
+  }
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/CommandObjects.java:657]: Do these in Jedis.
```suggestion
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**